### ISSUE TITLE
Ganeti KVM instance doesn't start with host-based kernel - KVM reports "Two devices with same boot index 0"

### ISSUE THREAD

**Issue by Ganeti-Issues-Migrator** [ganeti/ganeti#286]
Originally reported of Google Code with ID 230.

```
What software version are you running? Please provide the output of "gnt-
cluster --version" and "gnt-cluster version".


# gnt-cluster --version
gnt-cluster (ganeti v2.4.5) 2.4.5
# gnt-cluster version
Software version: 2.4.5
Internode protocol: 2040000
Configuration format: 2040000
OS api version: 20
Export interface: 0


<b>What distribution are you using?</b>

RHEL6

<b>What steps will reproduce the problem?</b>
1. Install Ganeti 2.4.5 on RHEL6
2. Create a KVM instance using a host-based kernel (kernel_path pointing to a kernel in the host OS filesystem)
3. Attempt to start the instance

What is the expected output? 

Instance started

What do you see instead?

gnt-instance start gram5p7dev.canterbury.ac.nz
Waiting for job 41 for gram5p7dev.canterbury.ac.nz...
Job 41 for gram5p7dev.canterbury.ac.nz has failed: Failure: command execution error:
Could not start instance: Hypervisor error: Failed to start instance gram5p7dev.canterbury.ac.nz: exited with exit code 1 (Two devices with same boot index 0)


<b>Please provide any additional information below.</b>

This appears to be caused by Ganeti instructing KVM to setup two different devices as boot devices:

-boot c -drive file=/dev/VolumeGroup00/b1be0ad4-7969-4139-921b-b2c00e4df2e0.disk0,format=raw,if=virtio,boot=on

Manually dropping the "boot=on" from the drive parameters fixes it.

I've been able to force Ganeti to do this (drop "boot=on" from drive parameters) by modifying hv_kvm.py to boot_disk (and boot_cdrom, boot_floppy, boot_network) to true when HV kernel_path is blank.

Patch attached.  

Fixes the problem for me.  

Happy to contribute this to the project.

Cheers,
Vladimir

PS: Credits for dropping the "boot=on" flag go to 
http://comments.gmane.org/gmane.comp.emulators.qemu/94012


-- 
Vladimir Mencl, Ph.D.
E-Research Services and Systems Consultant
BlueFern Computing Services
University of Canterbury
Private Bag 4800
Christchurch 8140
New Zealand

http://www.bluefern.canterbury.ac.nz
mailto:vladimir.mencl@canterbury.ac.nz
Phone: +64 3 364 3012
Mobile: +64 21 997 352
Fax: +64 3 364 3002

```
Originally added on 2012-04-19 04:59:22 +0000 UTC.

Attachments: 
[ganeti-kvm-hv-boot-patch.diff](https://storage.googleapis.com/google-code-attachments/ganeti/issue-230/comment-0/ganeti-kvm-hv-boot-patch.diff)

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2012-07-19 14:59:48 +0000 UTC.

Added Labels: Type-Defect Hypervisor-KVM

**Comment by Ganeti-Issues-Migrator** [comment]
```
Hi,

This looks like a good fix; however, before accepting the patch, I'd really appreciate if you can sign our CLA per http://code.google.com/p/ganeti/wiki/DeveloperGuidelines. Thanks in advance!
```
Originally added on 2012-07-21 17:56:14 +0000 UTC.

Changed State: Accepted

**Comment by Ganeti-Issues-Migrator** [comment]
```
Hi, just signed electronically. Cheers, Vladimir
```
Originally added on 2012-07-23 00:30:24 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Hi again,

I was going to apply the patch, when I realised that we already skip boot=on in some certain cases. In any case, I can't reproduce it right now, with KVM 0.12 or 1.0, so I'm wondering, what actual KVM version does RHEL 6 have? Can you give the output of kvm --help and the kvm package version?

I suspect something strange here…
```
Originally added on 2012-07-23 12:15:07 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Hi,

The mail problem that this patch is addressing is that when booting a kernel + ramdisk from the host system, KVM would receive two boot disk configurations: one for the host OS kernel+ramdisk and one for the first hard-disk.

The original code for 2.4.5 has:

* lines 528..531 set one of boot_{disk,cdrom,floppy,network} to true based on the boot order
* for boot_disk, lines 558..562 add "-boot c" to kvm arguments and ",boot=on" to disk parameters if boot_disk is true.  (It skips the ",boot=on" for certain types of hard-disks - IDE).

What my patch is doing is changing the first block of code to set the respective boot_<device> variable to true ONLY if we are not booting a kernel from the host, effectively deactivating the second block of code when booting from the host.

Haven't checked if this has changed in newer versions of Ganeti.

KVM version for my system is:
"QEMU PC emulator version 0.12.1 (qemu-kvm-0.12.1.2), Copyright (c) 2003-2008 Fabrice Bellard"
(this is the first line of the output - full output attached as a file).


Hope this provides enough information.

Cheers,
Vladimir





```
Originally added on 2012-07-24 11:11:16 +0000 UTC.

Attachments: 
[kvm-version.txt](https://storage.googleapis.com/google-code-attachments/ganeti/issue-230/comment-5/kvm-version.txt)

**Comment by Ganeti-Issues-Migrator** [comment]
```
Hmm, I see. It means then that "-boot c" and "-kernel" don't work together; I thought that only "-kernel" and "-driver ...,boot=on" is restricted.

I've applied a modified version of your patch and this will be included in 2.6.0 and eventually 2.5.2 (if we make another 2.5 release).

Thanks!
```
Originally added on 2012-07-24 15:14:01 +0000 UTC.

Changed State: Started
Added to Milestone: Release2.6

**Comment by Ganeti-Issues-Migrator** [comment]
```
Happy to contribute.  

BTW, for the future, what would be the preferred way to send (similar small) contributions?  Email to http://groups.google.com/group/ganeti-devel ?  Or a ticket here combined with an email (linking to the ticket) ?
```
Originally added on 2012-07-24 22:26:27 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Email as produced by git-format-patch would be the ideal method. If that is not feasible (you need to configure git send-email, etc., but just attaching diffs to email is also OK. Don't forget the Signed-off-by tag, that is much appreciated.

Since this was released in 2.5.2 yesterday, I'm going to mark this a fixed. Thanks again!
```
Originally added on 2012-07-25 09:49:17 +0000 UTC.

Changed State: Fixed
Added to Milestone: Release2.5
Removed from Milestone: Release2.6

**Comment by Ganeti-Issues-Migrator** [comment]
```
Migrate old issues from 'fixed' to 'released'.
```
Originally added on 2012-10-22 09:16:17 +0000 UTC.

Changed State: Released
